core/templates/dev/head/collection_editor/CollectionEditor.js

```diff
@@ -28,6 +28,64 @@ oppia.constant(
 oppia.constant(
   'EXPLORATION_SUMMARY_DATA_URL_TEMPLATE', '/explorationsummarieshandler/data');
 
+// Sorts the node list for display in collection editor frontend. The list
+// starts with a startNode which contains no prerequisite skills. After that
+// nodes are added to the list if their prerequisite skills match the acquired
+// skills of the last node that was added to the node list.
+oppia.filter('customSorter', function() {
+  return function(nodes) {
+    // Get starting node
+    var startingNode;
+    for (var i = 0; i < nodes.length; i++) {
+      if (nodes[i].getPrerequisiteSkillList().isEmpty()) {
+        startingNode = nodes[i];
+        break;
+      }
+    }
+
+    // Populate node list
+    var nodeList = [];
+    nodeList.push(startingNode);
+    while (true) {
```

Comment:
An alternative algorithm might be the following:

1. Start with a "skill pool" of empty array; this represents the skills currently unlocked
2. "Play" the first exploration by adding the skills acquired from it to the skill pool
3. Find the next exploration whose prerequisite skill list is a subset of the current skill pool
4. "Play" that exploration by adding its acquired skill list to the current skill pool
5. Repeat this until no more explorations can be found
6. The filter should return the collection nodes in the order they were encountered in the algorithm (any unreachable collection nodes not encountered should be ignored; these would cause an infinite loop in the current implementation)

See the [validation check](https://github.com/amgowano/oppia/blob/1ecb7f7198a3ae8705199a486ddbe7dc53cfba69/core/templates/dev/head/domain/collection/CollectionValidationService.js#L43) which does a similar approach of simulating a user playing through a collection, as it is similar to what's proposed above. It's up to you whether to change the filter, but I think the above is conceptually simpler. It's also very similar to what you're trying to do here, I think.